src/QMCDrivers/WFOpt/EngineHandle.h

```diff
@@ -0,0 +1,139 @@
+//////////////////////////////////////////////////////////////////////////////////////
+// This file is distributed under the University of Illinois/NCSA Open Source License.
+// See LICENSE file in top directory for details.
+//
+// Copyright (c) 2022 QMCPACK developers.
+//
+// File developed by: Ye Luo, yeluo@anl.gov, Argonne National Laboratory
+//                    Leon Otis, leon_otis@berkeley.edu, UC Berkeley
+//
+// File created by: Ye Luo, yeluo@anl.gov, Argonne National Laboratory
+//////////////////////////////////////////////////////////////////////////////////////
+
+#ifndef QMCPLUSPLUS_ENGINE_HANDLE_HEADER
+#define QMCPLUSPLUS_ENGINE_HANDLE_HEADER
+
+#include "Containers/MinimalContainers/RecordArray.hpp"
+#include "Concurrency/OpenMP.h"
+
+#include "QMCDrivers/Optimizers/DescentEngine.h"
+
+#ifdef HAVE_LMY_ENGINE
+#include "formic/utils/lmyengine/engine.h"
+#endif
+
+
+namespace qmcplusplus
+{
+class EngineHandle
+{
+public:
+  using Real          = QMCTraits::RealType;
+  using Value         = QMCTraits::ValueType;
+  using FullPrecReal  = QMCTraits::FullPrecRealType;
+  using FullPrecValue = QMCTraits::FullPrecValueType;
+
+  virtual void prepareSampling(int num_params) = 0;
+  virtual void takeSample(const std::vector<FullPrecReal>& energy_list,
+                          const RecordArray<Value>& dlogpsi_array,
+                          const RecordArray<Value>& dhpsioverpsi_array,
+                          int ib)              = 0;
+  virtual void finishSampling()                = 0;
```

Comment:
Please document all three APIs.

Reply:
Thanks @ye-luo for fixing the builds. I've added some documentation and cleaned up some commented code, but let me know if there's more that should be changed. My intent for the next PR after this one with the interface is done is to add the remaining changes in #4061 for the optimizers themselves. Changes would be inside the LM engine and QMCFixedSampleLinearOptimizeBatched.